File: backend/routers/tasks.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from dependencies import get_db, verify_api_key
from services.orchestrator import Orchestrator
from schemas.task import TaskCreate

router = APIRouter(tags=["Tasks"])


@router.post("/tasks", dependencies=[Depends(verify_api_key)])
async def create_task(body: TaskCreate, background_tasks: BackgroundTasks, db: AsyncSession = Depends(get_db)):
    orch = Orchestrator(db)
    result = await orch.create_task(body.session_id, body.type, body.description, body.requirements)
    background_tasks.add_task(Orchestrator(db).execute_task, result["task_id"])
    return result


@router.get("/tasks/{task_id}", dependencies=[Depends(verify_api_key)])
async def get_task(task_id: str, db: AsyncSession = Depends(get_db)):
    orch = Orchestrator(db)
    result = await orch.get_status(task_id)
    if not result:
        raise HTTPException(status_code=404, detail="Task not found")
    return result


@router.post("/tasks/{task_id}/cancel", dependencies=[Depends(verify_api_key)])
async def cancel_task(task_id: str, db: AsyncSession = Depends(get_db)):
    orch = Orchestrator(db)
    ok = await orch.cancel_task(task_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Task not found")
    return {"success": True, "message": "Task cancelled"}

File: frontend/src/api/tasks.ts
import { api } from './client';

export async function createTask(type: string, description: string, requirements: string[] = []) {
  return api.post<{ task_id: string; status: string; created_at: string }>('/tasks', { type, description, requirements });
}

export async function getTask(taskId: string) {
  return api.get<{ task_id: string; status: string; progress: number; sub_tasks: Array<{ name: string; status: string }>; results?: Record<string, string> }>(`/tasks/${taskId}`);
}

export async function cancelTask(taskId: string) {
  return api.post<{ success: boolean; message: string }>(`/tasks/${taskId}/cancel`);
}
